oio/directory/indexer.py

```diff
@@ -0,0 +1,257 @@
+                a default size of 10 will be created.
+        """
+        self.logger = get_logger(conf)
+        self.stop = False
+        self.volume = volume_path
+        self.success_nb = 0
+        self.failed_nb = 0
+        self.full_scan_nb = 0
+        self.last_report_time = 0
+        self.last_scan_time = 0
+        self.last_index_time = 0
+        self.start_time = 0
+        self.indexed_since_last_report = 0
+        self.scans_interval = int_value(
+            conf.get('interval'), 1800)
+        self.report_interval = int_value(
+            conf.get('report_interval'), 300)
+        self.max_indexed_per_second = int_value(
+            conf.get('scanned_per_second'), 3000)
+        self.namespace, self.volume_id = check_volume_for_service_type(
+            self.volume, "meta2")
+        if not pm:
+            pm = get_pool_manager(pool_connections=10)
+        self.index_client = RdirClient(conf, logger=self.logger,
+                                       pool_manager=pm)
+        self.dir_client = DirectoryClient(conf, logger=self.logger,
+                                          pool_manager=pm)
+
+    def report(self, tag):
+        """
+        Log the status of indexer
+
+        :param tag: One of three: starting, running, ended.
+        """
+        total = self.success_nb + self.failed_nb
+        now = time.time()
+        elapsed = (now - self.start_time) or 0.00001
+        since_last_rprt = (now - self.last_report_time) or 0.00001
+        self.logger.info(
+            'volume_id=%(volume_id)s %(tag)s=%(current_time)s '
+            'elapsed=%(elapsed).02f '
+            'pass=%(pass)d '
+            'errors=%(errors)d '
+            'containers_indexed=%(total_indexed)d %(index_rate).2f/s' % {
+                'volume_id': self.volume_id,
+                'tag': tag,
+                'current_time': datetime.fromtimestamp(
+                    int(now)).isoformat(),
+                'pass': self.full_scan_nb,
+                'errors': self.failed_nb,
+                'total_indexed': total,
+                'index_rate': self.indexed_since_last_report / since_last_rprt,
+                'elapsed': elapsed
+            }
+        )
+        self.last_report_time = now
+        self.indexed_since_last_report = 0
+
+    def warn(self, msg, container_id):
+        self.logger.warn(
+            'volume_id=%(volume_id)s container_id=%(container_id)s %(error)s' %
+            {
+                'volume_id': self.volume_id,
+                'container_id': container_id,
+                'error': msg
+            }
+        )
+
+    def index_meta2_database(self, db_id):
+        """
+        Add a meta2 database to the rdir index. Fails if the database isn't
+        handled by the current volume.
+
+        :param db_id: The ContentID representing the reference to the database.
+        """
+        if len(db_id) != STRLEN_REFERENCEID:
```

Comment:
This check will fail when the sequence number of the database will be greater than 9.